app/models/form/import.rb

```diff
@@ -151,14 +151,29 @@ def validate_data
     errors.add(:data, I18n.t('imports.errors.over_rows_processing_limit', count: ROWS_PROCESSING_LIMIT)) if csv_row_count > ROWS_PROCESSING_LIMIT
 
     if type.to_sym == :following
-      base_limit = FollowLimitValidator.limit_for_account(current_account)
-      limit = base_limit
-      limit -= current_account.following_count unless overwrite
-      errors.add(:data, I18n.t('users.follow_limit_reached', limit: base_limit)) if csv_row_count > limit
+      follows_limit = FollowLimitValidator.limit_for_account(current_account)
+      errors.add(:data, I18n.t('users.follow_limit_reached', limit: follows_limit)) if follows_count_after_import > follows_limit
     end
   rescue CSV::MalformedCSVError => e
     errors.add(:data, I18n.t('imports.errors.invalid_csv_file', error: e.message))
   rescue EmptyFileError
     errors.add(:data, I18n.t('imports.errors.empty'))
   end
+
+  def follows_count_after_import
+    return csv_row_count if overwrite
+
+    accts = parsed_rows.pluck('acct').compact.map(&:downcase).uniq.map do |acct|
+      username, domain = acct.split('@')
+      if domain.nil? || TagManager.instance.local_domain?(domain)
+        username.downcase
+      else
+        "#{username.downcase}@#{TagManager.instance.normalize_domain(domain)}"
+      end
+    end
+
+    current_follows = current_account.following.select(:username, :domain).map { |account| account.acct.downcase }
```

Comment:
This pulls the `username` and `domain` of *all* the accounts the current account follows in one SQL query. In most cases, this should be fine, but for large accounts, this could be a lot. Unfortunately I'm not too sure how to do differently.

The best alternative I have in mind is to have something like `WHERE NOT (lower(username), lower(domain)) IN (…)` but there are multiple issues with that:
- this could still be up to 20k items in the SQL query
- Rails offers no facility to do that efficiently

Reply:
> Before I suggest ideas here I want to make sure I understand the issue -- is this what's happening...
> 
>     * Goal is to arrive at a unique count of about-to-be-imported acct values, after we remove the ones the account is already following

The goal is to get the count of unique followed accounts if we imported all the accounts from the CSV.

>     * We are loading some CSV data so we wind up with a ruby data structure that has some mix of username, domain, acct value, etc - at our choosing, from the inbound file

It contains `acct` but they may need some normalization.

>     * We have to pull out a collection of all the username/domain/acct values from the follows of the account via sql

That, or as you said in your next query, have the SQL query include all the values from the CSV, which may be preferable depending on its size.

>     * This is an issue because we either write a massive sql query with all the values from the CSV, or we obtain all the values via sql and then do some ruby looping/comparing to get the uniques, and neither of these are great

Yeah, existing follows are theoretically unlimited (but in practice often under 10k) and accts to be imported are at most 20k, which is still a large value.

Reply:
Got it, and yeah neither option is appealing. Thoughts:

- Could we update the validation error message to be more deliberate about what was measured and that users might be able to self-correct by pruning their file of existing followers first? This I guess is technically an improvement and that it adds clarify ... but is well short of a good solution, since from the users perspective obviously the server knows what the list is and de-duping a list seems like a thing you shouldnt have to do yourself.
- Could we shift the whole thing to a background task, take advantage of more batching and slower loading, send users an email summary at the end of how many were dupes/new/failures/whatever? Could even special case this to have a background flow only if the rows in the file exceed some number? Do inline processing via either sql query or ruby looping inline in controller action when under that limit, go to background with email summary when over?

Reply:
The import process is already a background job with error tracking, but the limit check is done immediately so the users know why they can't import more followers… I guess we could remove the check altogether, or change it to only error out if the user has *already* reached their max follow count.

Reply:
Yeah, I meant just in the narrow upload/response interaction. The current error people are seeing has scenarios where it might be an inaccurate error, due to the counting method. Changing this to either not have a validation/warning at all, or so that instead of a validation error its a flash warning message -- and then also sending an email summary of what was imported ... might be better, if not perfect.